pkg/graph/builder.go

```diff
@@ -269,42 +300,105 @@ func (b *Builder) buildExternalRefResource(
 // buildRGResource builds a resource from the given resource definition.
 // It provides a high-level understanding of the resource, by extracting the
 // OpenAPI schema, emulating the resource and extracting the cel expressions
-// from the schema.
+// from the schema. Returns multiple errors if validation fails at different stages.
 func (b *Builder) buildRGResource(
 	rgResource *v1alpha1.Resource,
 	namespacedResources map[k8sschema.GroupKind]bool,
 	order int,
 ) (*Resource, error) {
+	// Collect all errors instead of returning early
+	errors := []error{}
+
 	// 1. We need to unmarshal the resource into a map[string]interface{} to
 	//    make it easier to work with.
 	resourceObject := map[string]interface{}{}
 	if len(rgResource.Template.Raw) > 0 {
 		err := yaml.UnmarshalStrict(rgResource.Template.Raw, &resourceObject)
 		if err != nil {
-			return nil, fmt.Errorf("failed to unmarshal resource %s: %w", rgResource.ID, err)
+			errors = append(errors,
+				NewValidationError(
+					ValidationErrorTypeResource,
+					rgResource.ID,
+					"",
+					fmt.Sprintf("resources.%s", rgResource.ID),
+					fmt.Errorf("failed to unmarshal resource template: %w", err),
+				),
+			)
 		}
 	} else if rgResource.ExternalRef != nil {
 		resourceObject = b.buildExternalRefResource(rgResource.ExternalRef)
 	} else {
-		return nil, fmt.Errorf("exactly one of template or externalRef must be provided")
+		errors = append(errors,
+			NewValidationError(
+				ValidationErrorTypeResource,
+				rgResource.ID,
+				"",
+				fmt.Sprintf("resources.%s", rgResource.ID),
+				fmt.Errorf("exactly one of template or externalRef must be provided"),
+			),
+		)
+	}
+
+	// If we couldn't get a valid resource object, return early with errors
+	if len(errors) > 0 {
+		return nil, &MultiError{Errors: errors}
```

Comment:
idk about these error returns (and a few below here) becausee you can basically just return it immediately in the error branch instead of doing this length check, because up until this point you cannot have multiple errors at once